src/green.rs

```diff
@@ -58,6 +62,31 @@ impl<T: Types> GreenNodeBuilder<T> {
         self.children
             .push(GreenNode::new_branch(kind, children.into_boxed_slice()));
     }
+    /// Prepare for maybe wrapping the next node.
+    /// The way wrapping works is that you first of all get a checkpoint,
+    /// then you place all tokens you want to wrap, and then *maybe* call wrap_internal.
+    /// Example:
+    /// ```rust,ignore
+    /// let checkpoint = builder.wrap_checkpoint();
+    /// self.parse_expr();
+    /// if self.peek() == Some(Token::Plus) {
+    ///   // 1 + 2 = Add(1, 2)
+    ///   builder.wrap_internal(checkpoint, Token::Operation);
+    ///   self.parse_expr();
+    ///   builder.finish_internal();
+    /// }
+    /// ```
+    pub fn wrap_checkpoint(&self) -> WrapCheckpoint {
+        WrapCheckpoint(self.children.len())
+    }
+    /// Wrap the previous branch marked by wrap_checkpoint in a new branch and
+    /// make it current.
+    pub fn wrap_internal(&mut self, checkpoint: WrapCheckpoint, kind: T::Kind) {
```

Comment:
Nifty!

I thought one needs to create a wrapper around green node to track such state (I do this in rust-analyzer), but looks like this indeed work!

Lets name it `start_internal_at(checkpoint, kind)`, to make it more obvious that `start _internal + start_internal_at = finish_internal`.

Also, I think there's one more invariant we need to check: the `fist_child` fields of `parents` should be increasing.

So, we need to check that top's first_child is <= checkpoint. I think? Not sure about this one.


It would be awesome to exercise this by writing a tests which builds a tree for `1 + 2 * 3  + 4` (without the parser, just calling the builder).

Reply:
I'm using this in rnix currently and it seems to work